internal/statestore/redis.go

```diff
@@ -56,65 +56,126 @@ func (rb *redisBackend) Close() error {
-	if len(passwordFile) > 0 {
-		redisLogger.Debugf("loading Redis password from file %s", passwordFile)
-		passwordData, err := ioutil.ReadFile(passwordFile)
-		if err != nil {
-			redisLogger.Fatalf("cannot read Redis password from file %s, desc: %s", passwordFile, err.Error())
+func getRedisPool(poolType string, cfg config.View) *redis.Pool {
+	var maxIdle, maxActive int
+	var idleTimeout time.Duration
+	var sentinelPool *redis.Pool
+
+	if cfg.GetBool("redis.sentinelEnabled") {
+		sentinelAddr := fmt.Sprintf("%s:%s", cfg.GetString("redis.sentinelHostname"), cfg.GetString("redis.sentinelPort"))
+		sentinelURL := redisURLFromAddr(sentinelAddr, cfg)
+		sentinelPool = &redis.Pool{
+			MaxIdle:      maxIdle,
+			MaxActive:    maxActive,
+			IdleTimeout:  idleTimeout,
+			Wait:         true,
+			TestOnBorrow: testOnBorrow,
+			DialContext: func(ctx context.Context) (redis.Conn, error) {
+				if ctx.Err() != nil {
+					return nil, ctx.Err()
+				}
+				redisLogger.WithField("sentinelAddr", sentinelAddr).Debug("Attempting to connect to Redis Sentinel")
+				return redis.DialURL(sentinelURL, redis.DialConnectTimeout(idleTimeout), redis.DialReadTimeout(idleTimeout))
+			},
 		}
-		redisURL += fmt.Sprintf("%s:%s@", cfg.GetString("redis.user"), string(passwordData))
-		maskedURL += fmt.Sprintf("%s:%s@", cfg.GetString("redis.user"), "**********")
 	}
-	redisURL += cfg.GetString("redis.hostname") + ":" + cfg.GetString("redis.port")
-	maskedURL += cfg.GetString("redis.hostname") + ":" + cfg.GetString("redis.port")
 
-	redisLogger.WithField("redisURL", maskedURL).Debug("Attempting to connect to Redis")
+	switch poolType {
+	case "healthcheck":
+		maxIdle = 3
+		maxActive = 0
+		idleTimeout = 10 * cfg.GetDuration("redis.pool.healthCheckTimeout")
+	default:
+		maxIdle = cfg.GetInt("redis.pool.maxIdle")
+		maxActive = cfg.GetInt("redis.pool.maxActive")
+		idleTimeout = cfg.GetDuration("redis.pool.idleTimeout")
+	}
 
-	pool := &redis.Pool{
-		MaxIdle:     cfg.GetInt("redis.pool.maxIdle"),
-		MaxActive:   cfg.GetInt("redis.pool.maxActive"),
-		IdleTimeout: cfg.GetDuration("redis.pool.idleTimeout"),
-		Wait:        true,
-		TestOnBorrow: func(c redis.Conn, lastUsed time.Time) error {
-			if time.Since(lastUsed) < 15*time.Second {
-				return nil
-			}
+	return &redis.Pool{
+		MaxIdle:      maxIdle,
+		MaxActive:    maxActive,
+		IdleTimeout:  idleTimeout,
+		Wait:         true,
+		TestOnBorrow: testOnBorrow,
+		DialContext:  getDialFunc(poolType, cfg, sentinelPool),
+	}
+}
 
-			_, err := c.Do("PING")
-			return err
-		},
-		DialContext: func(ctx context.Context) (redis.Conn, error) {
-			if ctx.Err() != nil {
-				return nil, ctx.Err()
+func getDialFunc(poolType string, cfg config.View, sentinelPool *redis.Pool) func(ctx context.Context) (redis.Conn, error) {
+	idleTimeout := cfg.GetDuration("redis.pool.idleTimeout")
+	healthCheckTimeout := cfg.GetDuration("redis.pool.healthCheckTimeout")
+
+	if cfg.GetBool("redis.sentinelEnabled") {
```

Comment:
if sentinelPool == nil? rather than looking up enabled? else for any reason if it is nil, you panic accessing it further below?